storage/src/checker/range_set.rs

```diff
@@ -0,0 +1,169 @@
+    type Item = (LinearAddress, u64);
+    type IntoIter = std::iter::Map<
+        <RangeSet<u64> as IntoIterator>::IntoIter,
+        fn(AnyRange<u64>) -> (LinearAddress, u64),
+    >;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.range_set.into_iter().map(|range| {
+            let Bound::Included(start) = range.start() else {
+                panic!("start of range is not included, this should not happen");
+            };
+            let Bound::Excluded(end) = range.end() else {
+                panic!("end of range is not excluded, this should not happen");
+            };
+            let start_addr =
+                NonZeroU64::new(*start).expect("found zero address, this should not happen");
+            (start_addr, end - start)
+        })
+    }
+}
+
+#[cfg(test)]
+#[expect(clippy::unwrap_used)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn test_empty() {
+        let visited = LinearAddressRangeSet::new(0x1000).unwrap();
+        assert_eq!(visited.into_iter().collect::<Vec<_>>(), vec![]);
+    }
+
+    #[test]
+    fn test_insert_area() {
+        let mut visited = LinearAddressRangeSet::new(0x1000).unwrap();
+        let start = LinearAddress::new(2048).unwrap();
+        let size = 1024;
+
+        visited
+            .insert_area(start, size)
+            .expect("the given area should be within bounds");
+
+        assert_eq!(visited.into_iter().collect::<Vec<_>>(), vec![(start, size)]);
+    }
+
+    #[test]
+    fn test_consecutive_areas_merge() {
+        let mut visited = LinearAddressRangeSet::new(0x1000).unwrap();
+        let start1 = 2048;
+        let size1 = 1024;
+        let start2 = start1 + size1;
+        let size2 = 1024;
+
+        let start1_addr = LinearAddress::new(start1).unwrap();
+        let start2_addr = LinearAddress::new(start2).unwrap();
+
+        visited
+            .insert_area(start1_addr, size1)
+            .expect("the given area should be within bounds");
+
+        visited
+            .insert_area(start2_addr, size2)
+            .expect("the given area should be within bounds");
+
+        assert_eq!(
+            visited.into_iter().collect::<Vec<_>>(),
+            vec![(start1_addr, size1 + size2)]
+        );
+    }
+
+    #[test]
+    fn test_intersecting_areas_will_fail() {
+        let start1 = 2048;
+        let size1 = 1024;
+        let start2 = start1 + size1 - 1;
+        let size2 = 1024;
+
+        let start1_addr = LinearAddress::new(start1).unwrap();
+        let start2_addr = LinearAddress::new(start2).unwrap();
+
+        let mut visited = LinearAddressRangeSet::new(0x1000).unwrap();
+        visited
+            .insert_area(start1_addr, size1)
+            .expect("the given area should be within bounds");
+
+        let error = visited
+            .insert_area(start2_addr, size2)
+            .expect_err("the given area should intersect with the first area");
+
+        let CheckerError::AreaIntersects {
```

Comment:
These are easier to write with assert_matches

https://doc.rust-lang.org/std/assert_matches/macro.assert_matches.html

Reply:
Yes, but you can still do it with:

```rust
assert!(matches!(...))
```

Reply:
I don't think `matches!` also lets you check the values inside? Part of reason I am doing this now is because I want to take the values in the error type out and check they are correct next.